environs/config/validator.go

```diff
@@ -13,3 +25,110 @@ type Validator interface {
 	// use a separate mechanism for updating config.
 	Validate(cfg, old *Config) (valid *Config, _ error)
 }
+
+// ValidatorFunc is utility type for declaring func's that implement the
+// Validator interface.
+type ValidatorFunc func(cfg, old *Config) (*Config, error)
+
+// ValidationError represents a specific error that has occurred when validating
+// Config. It allows for the placement of one or more attributes with a reason
+// as to why either their keys or values are not valid in the current context.
+type ValidationError struct {
+	// InvalidAttrs is the list of attributes in the config that where invalid.
+	InvalidAttrs []string
+
+	// Reason is the human readable message for why the attributes are invalid.
+	Reason string
+}
+
+var (
+	DisallowedModelConfigAttrs = [...]string{
+		AgentVersionKey,
+		AdminSecretKey,
+		CAPrivateKeyKey,
+	}
+)
+
+// Is implements errors Is interface. We implement is so that the
+// ValidationError also satisfies NotValid.
+func (v *ValidationError) Is(err error) bool {
+	return err == errors.NotValid || err == v
+}
+
+// Error implements Error interface.
+func (v *ValidationError) Error() string {
+	return fmt.Sprintf("config attributes %v not valid because %s", v.InvalidAttrs, v.Reason)
+}
+
+// Validate implements Validator validate interface. This func will run all the
+// validators in the aggregate till either a validator errors or there are no
+// more validators to run. The returned config from each validator is passed
+// into the subsequent validator.
+func (a *AggregateValidator) Validate(cfg, old *Config) (*Config, error) {
+	var err error
+	for i, validator := range a.Validators {
+		cfg, err = validator.Validate(cfg, old)
+		if err != nil {
+			return cfg, fmt.Errorf("config validator %d failed: %w", i, err)
+		}
+	}
+	return cfg, nil
+}
+
+// Validate implements the Validator interface.
+func (v ValidatorFunc) Validate(cfg, old *Config) (*Config, error) {
+	return v(cfg, old)
+}
+
+// ControllerNotContainingValidator implements a validator that asserts if the
+// supplied config contains and controller specific configuration attributes. A
+// ValidationError is returned if the config contains controller attributes.
+func ControllerNotContainingValidator() Validator {
```

Comment:
Maybe call it "NoControllerAttributesValidator". The current name seems yukky to me :-)